(Serious?) warnings in unittests for ATmega

Running

     scons unittest target=atmega

results in

     Warn: Specifying 'general.name' is deprecated and replaced by 'build.name'!

     xpcc/communication/sab/test/slave_test.cpp: In member function 'virtual void SlaveTest::tearDown()':
     xpcc/communication/sab/test/slave_test.cpp:123:9: warning: deleting object of polymorphic class type      'xpcc::sab::Slave<xpcc::sab::Interface<FakeIODevice> >' which has non-virtual destructor might cause      undefined behaviour [-Wdelete-non-virtual-dtor]
       delete slave;
              ^
     xpcc/communication/xpcc/test/dispatcher_test.cpp: In member function 'virtual void      DispatcherTest::tearDown()':
     xpcc/communication/xpcc/test/dispatcher_test.cpp:55:9: warning: deleting object of polymorphic class      type 'TestingComponent2' which has non-virtual destructor might cause undefined behaviour      [-Wdelete-non-virtual-dtor]
       delete component2;
              ^
     xpcc/communication/xpcc/test/dispatcher_test.cpp:56:9: warning: deleting object of polymorphic class      type 'TestingComponent1' which has non-virtual destructor might cause undefined behaviour      [-Wdelete-non-virtual-dtor]
       delete component1;
              ^
     xpcc/communication/xpcc/test/dispatcher_test.cpp:59:9: warning: deleting object of polymorphic class      type 'FakePostman' which has non-virtual destructor might cause undefined behaviour      [-Wdelete-non-virtual-dtor]
       delete postman;
              ^
     In file included from ./unittest/testsuite.hpp:34:0,
                      from xpcc/communication/xpcc/test/dispatcher_test.hpp:34,
                      from xpcc/communication/xpcc/test/dispatcher_test.cpp:32:
     ./unittest/harness.hpp: In instantiation of 'bool unittest::checkEqual(const A&, const B&, unsigned      int) [with A = unsigned int; B = int]':
     xpcc/communication/xpcc/test/dispatcher_test.cpp:169:2:   required from here
     ./unittest/harness.hpp:135:9: warning: comparison between signed and unsigned integer expressions      [-Wsign-compare]
        if (a == b) {
              ^

     In file included from ./unittest/testsuite.hpp:34:0,
                      from xpcc/io/test/io_stream_test.hpp:10,
                      from xpcc/io/test/io_stream_test.cpp:10:
     ./unittest/harness.hpp: In instantiation of 'bool unittest::checkEqual(const A&, const B&, unsigned      int) [with A = int; B = unsigned int]':
     xpcc/io/test/io_stream_test.cpp:559:2:   required from here
     ./unittest/harness.hpp:135:9: warning: comparison between signed and unsigned integer expressions      [-Wsign-compare]
        if (a == b) {
              ^
     In file included from ./xpcc/driver/pressure/bme280_data.hpp:236:0,
                      from ./xpcc/driver/pressure/bme280.hpp:18,
                      from xpcc/driver/pressure/test/bme280_test.cpp:1:
     ./xpcc/driver/pressure/bme280_data_impl_fp.hpp: In member function 'void      xpcc::bme280data::Data::calculateCalibratedTemperature()':
     ./xpcc/driver/pressure/bme280_data_impl_fp.hpp:31:28: warning: left shift count >= width of type
       int32_t adc = ((raw[3] << 16) | (raw[4] << 8) | (raw[5] << 0));
                                 ^
     ./xpcc/driver/pressure/bme280_data_impl_fp.hpp: In member function 'void      xpcc::bme280data::Data::calculateCalibratedPressure()':
     ./xpcc/driver/pressure/bme280_data_impl_fp.hpp:64:28: warning: left shift count >= width of type
       int32_t adc = ((raw[0] << 16) | (raw[1] << 8) | (raw[2] << 0));
                                 ^
     In file included from ./xpcc/driver/pressure/bme280_data.hpp:237:0,
                      from ./xpcc/driver/pressure/bme280.hpp:18,
                      from xpcc/driver/pressure/test/bme280_test.cpp:1:
     ./xpcc/driver/pressure/bme280_data_impl_double.hpp: In member function 'void      xpcc::bme280data::DataDouble::calculateCalibratedTemperature()':
     ./xpcc/driver/pressure/bme280_data_impl_double.hpp:33:28: warning: left shift count >= width of type
       int32_t adc = ((raw[3] << 16) | (raw[4] << 8) | (raw[5] << 0));
                                 ^
     ./xpcc/driver/pressure/bme280_data_impl_double.hpp: In member function 'void      xpcc::bme280data::DataDouble::calculateCalibratedPressure()':
     ./xpcc/driver/pressure/bme280_data_impl_double.hpp:65:28: warning: left shift count >= width of type
       int32_t adc = ((raw[0] << 16) | (raw[1] << 8) | (raw[2] << 0));
                                 ^
     Compiling C++: /Users/usr/Dev/2017amt/xpcc_feature_rtc_ds1302_support/build/unittest_atmega/xpcc/math/     filter/test/debounce_test.o
     xpcc/driver/pressure/test/bmp085_test.cpp: In member function 'void Bmp085Test::testConversion()':
     xpcc/driver/pressure/test/bmp085_test.cpp:117:14: warning: comparison between signed and unsigned      integer expressions [-Wsign-compare]
         if (error > max_error) {
                   ^
     xpcc/driver/pressure/test/bmp085_test.cpp:182:15: warning: comparison between signed and unsigned      integer expressions [-Wsign-compare]
          if (error > max_error) {
                    ^

The current unittests can only by run on an ATMega2560 with `gcc version 4.9.3 (GCC)`. Others do not have enough flash.

     Device: atmega2560

     Program:  242610 bytes (92.5% Full)
     (.text + .data + .bootloader)

I will try to work on the BME/BMP issues. What about the other warnings?

--- a/scons/site_tools/avr.py
+++ b/scons/site_tools/avr.py
@@ -136,6 +136,11 @@ def generate(env, **kw):
 		"-Wall",
 		"-Wextra",
 		"-Wundef",
+#		If we had a newer avr-g++ :-(
+#		"-Werror=shift-count-negative",
+#		"-Werror=shift-count-overflow",
+#		"-Werror=shift-negative-value",
+#		"-Werror=shift-overflow",
 #		"-Winline",
 		"-Wa,-adhlns=${TARGET.base}.lst",
 		"-DBASENAME=${SOURCE.file}",

--- a/src/unittest_atmega.cfg
+++ b/src/unittest_atmega.cfg
@@ -1,8 +1,6 @@
-[general]
+[build]
 name = unittest_atmega
 unittest = true
-
-[build]
 device = atmega2560
 clock = 1000000
 template = ../templates/unittest/runner_atmega.cpp.in

--- a/src/unittest_atxmega.cfg
+++ b/src/unittest_atxmega.cfg
@@ -1,9 +1,6 @@
-[general]
+[build]
 name = unittest_atxmega
 unittest = true
-
-[build]
-architecture = avr
 device = atxmega128a1
 clock = 2000000
 template = ../templates/unittest/runner_atxmega.cpp.in

--- a/src/unittest_hosted.cfg
+++ b/src/unittest_hosted.cfg
@@ -1,7 +1,5 @@
-[general]
-unittest = true
-
 [build]
+unittest = true
 device = hosted
 template = ../templates/unittest/runner_hosted.cpp.in
 buildpath = ../build/unittest_hosted

--- a/src/unittest_stm32.cfg
+++ b/src/unittest_stm32.cfg
@@ -1,7 +1,5 @@
-[general]
-unittest = true
-
 [build]
+unittest = true
 name = unittest_stm32
 board = stm32f4_discovery
 template = ../templates/unittest/runner_stm32.cpp.in

--- a/src/xpcc/communication/sab/slave.hpp
+++ b/src/xpcc/communication/sab/slave.hpp
@@ -49,6 +49,9 @@ namespace xpcc
 		class Transmitter
 		{
 		public:
+			virtual ~Transmitter()
+				{};
+
 			virtual void
 			send(bool acknowledge, const void *payload, std::size_t payloadLength) = 0;
 		};

--- a/src/xpcc/communication/xpcc/test/dispatcher_test.cpp
+++ b/src/xpcc/communication/xpcc/test/dispatcher_test.cpp
@@ -166,7 +166,7 @@ DispatcherTest::testEventTransmission()
 	TEST_ASSERT_TRUE(timeline->events.getFront().type == Timeline::EVENT);
 	TEST_ASSERT_EQUALS(timeline->events.getFront().id, 0x21);
 	TEST_ASSERT_EQUALS(timeline->events.getFront().source, 2);
-	TEST_ASSERT_EQUALS(timeline->events.getFront().payload.getSize(), 4);
+	TEST_ASSERT_EQUALS(timeline->events.getFront().payload.getSize(), 4U);
 	
 	// Event was also send to the backend (broadcast)
 	TEST_ASSERT_EQUALS(backend->messagesSend.getSize(), 1U);

--- a/src/xpcc/communication/xpcc/test/fake_postman.hpp
+++ b/src/xpcc/communication/xpcc/test/fake_postman.hpp
@@ -46,6 +46,9 @@
 class FakePostman : public xpcc::Postman
 {
 public:
+	virtual ~FakePostman()
+		{};
+
 	virtual DeliverInfo
 	deliverPacket(const xpcc::Header& header,
 			const xpcc::SmartPointer& payload);

--- a/src/xpcc/communication/xpcc/test/testing_component_1.hpp
+++ b/src/xpcc/communication/xpcc/test/testing_component_1.hpp
@@ -39,6 +39,9 @@ class TestingComponent1 : public xpcc::AbstractComponent
 {
 public:
 	TestingComponent1(xpcc::Dispatcher &communication, Timeline *timeline);
+
+	virtual ~TestingComponent1()
+		{};
 	
 	virtual void
 	update();

--- a/src/xpcc/communication/xpcc/test/testing_component_2.hpp
+++ b/src/xpcc/communication/xpcc/test/testing_component_2.hpp
@@ -39,6 +39,9 @@ class TestingComponent2 : public xpcc::AbstractComponent
 {
 public:
 	TestingComponent2(xpcc::Dispatcher &communication, Timeline *timeline);
+
+	virtual ~TestingComponent2()
+		{};
 	
 	virtual void
 	update();

--- a/src/xpcc/driver/pressure/bme280_data_impl_double.hpp
+++ b/src/xpcc/driver/pressure/bme280_data_impl_double.hpp
@@ -30,7 +30,7 @@ DataDouble::calculateCalibratedTemperature()
 {
 	// Temperature value is expected to be received in 20 bit format, 
 	// positive, stored in a 32 bit signed integer.
-	int32_t adc = ((raw[3] << 16) | (raw[4] << 8) | (raw[5] << 0));
+	int32_t adc = (((int32_t(raw[3])) << 16) | (raw[4] << 8) | (raw[5] << 0));
 	adc >>= 4;
 
 	double adc_16384 = double(adc) / double(16384.0);
@@ -62,7 +62,7 @@ DataDouble::calculateCalibratedPressure()
 		calculateCalibratedTemperature();
 	}
 
-	int32_t adc = ((raw[0] << 16) | (raw[1] << 8) | (raw[2] << 0));
+	int32_t adc = (((int32_t(raw[0])) << 16) | (raw[1] << 8) | (raw[2] << 0));
 	adc >>= 4;
 
 	double P1 = double(calibration.P1);

--- a/src/xpcc/driver/pressure/bme280_data_impl_fp.hpp
+++ b/src/xpcc/driver/pressure/bme280_data_impl_fp.hpp
@@ -28,7 +28,7 @@ namespace bme280data {
 void
 Data::calculateCalibratedTemperature()
 {
-	int32_t adc = ((raw[3] << 16) | (raw[4] << 8) | (raw[5] << 0));
+	int32_t adc = (((int32_t(raw[3])) << 16) | (raw[4] << 8) | (raw[5] << 0));
 	adc >>= 4;
 
 	XPCC_LOG_DEBUG.printf("adc = 0x%05x\n", adc);
@@ -61,7 +61,7 @@ Data::calculateCalibratedPressure()
 		calculateCalibratedTemperature();
 	}
 
-	int32_t adc = ((raw[0] << 16) | (raw[1] << 8) | (raw[2] << 0));
+	int32_t adc = (((int32_t(raw[0])) << 16) | (raw[1] << 8) | (raw[2] << 0));
 	adc >>= 4;
 
 	int64_t P1 = calibration.P1;

--- a/src/xpcc/driver/pressure/test/bmp085_test.cpp
+++ b/src/xpcc/driver/pressure/test/bmp085_test.cpp
@@ -112,10 +112,10 @@ Bmp085Test::testConversion()
 
 			XPCC_LOG_DEBUG.printf("%04x\t%5d\t%5d\t%2d\n", adc, tempDouble, temp, error);
 
-			error = std::abs(error);
-			total_error += error;
-			if (error > max_error) {
-				max_error = error;
+			uint16_t error_abs = std::abs(error);
+			total_error += error_abs;
+			if (error_abs > max_error) {
+				max_error = error_abs;
 			}
 		}
 
@@ -177,10 +177,10 @@ Bmp085Test::testConversion()
 				int16_t error = pressFp - pressDp;
 				XPCC_LOG_DEBUG.printf("  adc_press = %04x  PressFp = %6d Pa\t PressDp = %6d Pa \t Diff = %3d Pa\n", adc_press, pressFp, pressDp, error);
 
-				error = std::abs(error);
-				total_error += error;
-				if (error > max_error) {
-					max_error = error;
+				uint16_t error_abs = std::abs(error);
+				total_error += error_abs;
+				if (error_abs > max_error) {
+					max_error = error_abs;
 				}
 			}
 		}

--- a/src/xpcc/io/test/io_stream_test.cpp
+++ b/src/xpcc/io/test/io_stream_test.cpp
@@ -57,7 +57,7 @@ public:
 
 	static constexpr std::size_t buffer_length = 100;
 	char buffer[buffer_length];
-	int bytesWritten;
+	size_t bytesWritten;
 };
 
 // ----------------------------------------------------------------------------
@@ -85,7 +85,7 @@ IoStreamTest::testString()
 	(*stream) << string;
 
 	TEST_ASSERT_EQUALS_ARRAY(string, device.buffer, 6);
-	TEST_ASSERT_EQUALS(device.bytesWritten, 6);
+	TEST_ASSERT_EQUALS(device.bytesWritten, 6U);
 }
 
 FLASH_STORAGE_STRING(flashString) = "abc";
@@ -98,7 +98,7 @@ IoStreamTest::testFlashString()
 	(*stream) << xpcc::accessor::asFlash(flashString);
 
 	TEST_ASSERT_EQUALS_ARRAY(string, device.buffer, 3);
-	TEST_ASSERT_EQUALS(device.bytesWritten, 3);
+	TEST_ASSERT_EQUALS(device.bytesWritten, 3U);
 }
 
 // ----------------------------------------------------------------------------
@@ -110,7 +110,7 @@ IoStreamTest::testStreamUint8()
 	(*stream) << static_cast<uint8_t>(244);
 
 	TEST_ASSERT_EQUALS_ARRAY(string, device.buffer, 3);
-	TEST_ASSERT_EQUALS(device.bytesWritten, 3);
+	TEST_ASSERT_EQUALS(device.bytesWritten, 3U);
 }
 
 void
@@ -121,7 +121,7 @@ IoStreamTest::testStreamInt8()
 	(*stream) << static_cast<int8_t>(-123);
 
 	TEST_ASSERT_EQUALS_ARRAY(string, device.buffer, 4);
-	TEST_ASSERT_EQUALS(device.bytesWritten, 4);
+	TEST_ASSERT_EQUALS(device.bytesWritten, 4U);
 }
 
 // ----------------------------------------------------------------------------
@@ -133,7 +133,7 @@ IoStreamTest::testStreamUint16()
 	(*stream) << static_cast<uint16_t>(62412);
 
 	TEST_ASSERT_EQUALS_ARRAY(string, device.buffer, 5);
-	TEST_ASSERT_EQUALS(device.bytesWritten, 5);
+	TEST_ASSERT_EQUALS(device.bytesWritten, 5U);
 }
 
 void
@@ -144,7 +144,7 @@ IoStreamTest::testStreamUint16_2()
 	(*stream) << static_cast<uint16_t>(0xffff);
 
 	TEST_ASSERT_EQUALS_ARRAY(string, device.buffer, 5);
-	TEST_ASSERT_EQUALS(device.bytesWritten, 5);
+	TEST_ASSERT_EQUALS(device.bytesWritten, 5U);
 }
 
 // ----------------------------------------------------------------------------
@@ -156,7 +156,7 @@ IoStreamTest::testStreamInt16()
 	(*stream) << static_cast<int16_t>(-12345);
 
 	TEST_ASSERT_EQUALS_ARRAY(string, device.buffer, 6);
-	TEST_ASSERT_EQUALS(device.bytesWritten, 6);
+	TEST_ASSERT_EQUALS(device.bytesWritten, 6U);
 }
 
 void
@@ -167,7 +167,7 @@ IoStreamTest::testStreamInt16_2()
 	(*stream) << static_cast<int16_t>(-32768);
 
 	TEST_ASSERT_EQUALS_ARRAY(string, device.buffer, 6);
-	TEST_ASSERT_EQUALS(device.bytesWritten, 6);
+	TEST_ASSERT_EQUALS(device.bytesWritten, 6U);
 }
 
 // ----------------------------------------------------------------------------
@@ -179,7 +179,7 @@ IoStreamTest::testStreamUint32()
 	(*stream) << static_cast<uint32_t>(123);
 
 	TEST_ASSERT_EQUALS_ARRAY(string, device.buffer, 3);
-	TEST_ASSERT_EQUALS(device.bytesWritten, 3);
+	TEST_ASSERT_EQUALS(device.bytesWritten, 3U);
 }
 
 void
@@ -190,7 +190,7 @@ IoStreamTest::testStreamUint32_2()
 	(*stream) << static_cast<uint32_t>(0xffffffff);
 
 	TEST_ASSERT_EQUALS_ARRAY(string, device.buffer, 10);
-	TEST_ASSERT_EQUALS(device.bytesWritten, 10);
+	TEST_ASSERT_EQUALS(device.bytesWritten, 10U);
 }
 
 // ----------------------------------------------------------------------------
@@ -202,7 +202,7 @@ IoStreamTest::testStreamInt32()
 	(*stream) << static_cast<int32_t>(-12345678);
 
 	TEST_ASSERT_EQUALS_ARRAY(string, device.buffer, 9);
-	TEST_ASSERT_EQUALS(device.bytesWritten, 9);
+	TEST_ASSERT_EQUALS(device.bytesWritten, 9U);
 }
 
 void
@@ -214,7 +214,7 @@ IoStreamTest::testStreamInt32_2()
 	(*stream) << static_cast<int32_t>(-2147483647 - 1);
 
 	TEST_ASSERT_EQUALS_ARRAY(string, device.buffer, 11);
-	TEST_ASSERT_EQUALS(device.bytesWritten, 11);
+	TEST_ASSERT_EQUALS(device.bytesWritten, 11U);
 }
 
 void
@@ -225,7 +225,7 @@ IoStreamTest::testStreamInt32_3()
 	(*stream) << static_cast<int32_t>(0);
 
 	TEST_ASSERT_EQUALS_ARRAY(string, device.buffer, 1);
-	TEST_ASSERT_EQUALS(device.bytesWritten, 1);
+	TEST_ASSERT_EQUALS(device.bytesWritten, 1U);
 }
 
 // ----------------------------------------------------------------------------
@@ -238,7 +238,7 @@ IoStreamTest::testStreamUint64()
 	(*stream) << static_cast<uint64_t>(12345678901234ULL);
 
 	TEST_ASSERT_EQUALS_ARRAY(string, device.buffer, 14);
-	TEST_ASSERT_EQUALS(device.bytesWritten, 14);
+	TEST_ASSERT_EQUALS(device.bytesWritten, 14U);
 #endif
 }
 
@@ -251,7 +251,7 @@ IoStreamTest::testStreamInt64()
 	(*stream) << static_cast<int64_t>(-12345678901234LL);
 
 	TEST_ASSERT_EQUALS_ARRAY(string, device.buffer, 15);
-	TEST_ASSERT_EQUALS(device.bytesWritten, 15);
+	TEST_ASSERT_EQUALS(device.bytesWritten, 15U);
 #endif
 }
 
@@ -264,7 +264,7 @@ IoStreamTest::testFloat()
 	(*stream) << 1.23f;
 
 	TEST_ASSERT_EQUALS_ARRAY(string, device.buffer, 11);
-	TEST_ASSERT_EQUALS(device.bytesWritten, 11);
+	TEST_ASSERT_EQUALS(device.bytesWritten, 11U);
 }
 
 void
@@ -275,7 +275,7 @@ IoStreamTest::testFloat2()
 	(*stream) << 457.0f;
 
 	TEST_ASSERT_EQUALS_ARRAY(string, device.buffer, 11);
-	TEST_ASSERT_EQUALS(device.bytesWritten, 11);
+	TEST_ASSERT_EQUALS(device.bytesWritten, 11U);
 }
 
 void
@@ -286,7 +286,7 @@ IoStreamTest::testFloat3()
 	(*stream) << -51231400.0f;
 
 	TEST_ASSERT_EQUALS_ARRAY(string, device.buffer, 12);
-	TEST_ASSERT_EQUALS(device.bytesWritten, 12);
+	TEST_ASSERT_EQUALS(device.bytesWritten, 12U);
 }
 
 void
@@ -297,7 +297,7 @@ IoStreamTest::testFloat4()
 	(*stream) << -0.0007234f;
 
 	TEST_ASSERT_EQUALS_ARRAY(string, device.buffer, 12);
-	TEST_ASSERT_EQUALS(device.bytesWritten, 12);
+	TEST_ASSERT_EQUALS(device.bytesWritten, 12U);
 }
 
 void
@@ -310,7 +310,7 @@ IoStreamTest::testBool1()
 	(*stream) << boo;
 
 	TEST_ASSERT_EQUALS_ARRAY(string, device.buffer, 4);
-	TEST_ASSERT_EQUALS(device.bytesWritten, 4);
+	TEST_ASSERT_EQUALS(device.bytesWritten, 4U);
 }
 
 void
@@ -323,7 +323,7 @@ IoStreamTest::testBool2()
 	(*stream) << boo;
 
 	TEST_ASSERT_EQUALS_ARRAY(string, device.buffer, 5);
-	TEST_ASSERT_EQUALS(device.bytesWritten, 5);
+	TEST_ASSERT_EQUALS(device.bytesWritten, 5U);
 }
 
 // ----------------------------------------------------------------------------
@@ -337,7 +337,7 @@ IoStreamTest::testHex1()
 	(*stream) << xpcc::hex << c;
 
 	TEST_ASSERT_EQUALS_ARRAY(string, device.buffer, 2);
-	TEST_ASSERT_EQUALS(device.bytesWritten, 2);
+	TEST_ASSERT_EQUALS(device.bytesWritten, 2U);
 }
 
 void
@@ -350,7 +350,7 @@ IoStreamTest::testHex2()
 	(*stream) << xpcc::hex << s;
 
 	TEST_ASSERT_EQUALS_ARRAY(string, device.buffer, 10);
-	TEST_ASSERT_EQUALS(device.bytesWritten, 10);
+	TEST_ASSERT_EQUALS(device.bytesWritten, 10U);
 }
 
 void
@@ -363,7 +363,7 @@ IoStreamTest::testHex3()
 	(*stream) << xpcc::hex << i;
 
 	TEST_ASSERT_EQUALS_ARRAY(string, device.buffer, 4);
-	TEST_ASSERT_EQUALS(device.bytesWritten, 4);
+	TEST_ASSERT_EQUALS(device.bytesWritten, 4U);
 }
 
 void
@@ -376,7 +376,7 @@ IoStreamTest::testHex4()
 	(*stream) << xpcc::hex << i;
 
 	TEST_ASSERT_EQUALS_ARRAY(string, device.buffer, 8);
-	TEST_ASSERT_EQUALS(device.bytesWritten, 8);
+	TEST_ASSERT_EQUALS(device.bytesWritten, 8U);
 }
 
 void
@@ -392,7 +392,7 @@ IoStreamTest::testHex5()
 	(*stream) << xpcc::hex << boo;
 
 	TEST_ASSERT_EQUALS_ARRAY(string, device.buffer, 4);
-	TEST_ASSERT_EQUALS(device.bytesWritten, 4);
+	TEST_ASSERT_EQUALS(device.bytesWritten, 4U);
 }
 
 // ----------------------------------------------------------------------------
@@ -406,7 +406,7 @@ IoStreamTest::testBin1()
 	(*stream) << xpcc::bin << c;
 
 	TEST_ASSERT_EQUALS_ARRAY(string, device.buffer, 8);
-	TEST_ASSERT_EQUALS(device.bytesWritten, 8);
+	TEST_ASSERT_EQUALS(device.bytesWritten, 8U);
 }
 
 void
@@ -419,7 +419,7 @@ IoStreamTest::testBin2()
 	(*stream) << xpcc::bin << s;
 
 	TEST_ASSERT_EQUALS_ARRAY(string, device.buffer, 40);
-	TEST_ASSERT_EQUALS(device.bytesWritten, 40);
+	TEST_ASSERT_EQUALS(device.bytesWritten, 40U);
 }
 
 void
@@ -432,7 +432,7 @@ IoStreamTest::testBin3()
 	(*stream) << xpcc::bin << i;
 
 	TEST_ASSERT_EQUALS_ARRAY(string, device.buffer, 16);
-	TEST_ASSERT_EQUALS(device.bytesWritten, 16);
+	TEST_ASSERT_EQUALS(device.bytesWritten, 16U);
 }
 
 void
@@ -445,7 +445,7 @@ IoStreamTest::testBin4()
 	(*stream) << xpcc::bin << i;
 
 	TEST_ASSERT_EQUALS_ARRAY(string, device.buffer, 32);
-	TEST_ASSERT_EQUALS(device.bytesWritten, 32);
+	TEST_ASSERT_EQUALS(device.bytesWritten, 32U);
 }
 
 void
@@ -461,7 +461,7 @@ IoStreamTest::testBin5()
 	(*stream) << xpcc::bin << boo;
 
 	TEST_ASSERT_EQUALS_ARRAY(string, device.buffer, 16);
-	TEST_ASSERT_EQUALS(device.bytesWritten, 16);
+	TEST_ASSERT_EQUALS(device.bytesWritten, 16U);
 }
 
 void
@@ -471,7 +471,7 @@ IoStreamTest::testPrintf1()
 	(*stream).printf("Lala");
 
 	TEST_ASSERT_EQUALS_ARRAY("Lala", device.buffer, 4);
-	TEST_ASSERT_EQUALS(device.bytesWritten, 4);
+	TEST_ASSERT_EQUALS(device.bytesWritten, 4U);
 }
 
 void
@@ -505,7 +505,7 @@ IoStreamTest::testPrintf2()
 
 				char glibc[device.buffer_length];
 
-				int len = snprintf(glibc, device.buffer_length, fmt_str, ff);
+				size_t len = snprintf(glibc, device.buffer_length, fmt_str, ff);
 
 				(*stream).printf(fmt_str, ff);
 

